forge-ai/src/main/java/forge/ai/ability/DiscoverAi.java

```diff
@@ -0,0 +1,70 @@
+package forge.ai.ability;
+
+import forge.ai.AiPlayDecision;
+import forge.ai.ComputerUtil;
+import forge.ai.PlayerControllerAi;
+import forge.ai.SpellAbilityAi;
+import forge.card.CardStateName;
+import forge.game.ability.AbilityUtils;
+import forge.game.card.Card;
+import forge.game.player.Player;
+import forge.game.player.PlayerActionConfirmMode;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.Spell;
+import forge.game.spellability.SpellAbility;
+
+import java.util.Map;
+
+public class DiscoverAi extends SpellAbilityAi {
+
+    @Override
+    protected boolean checkApiLogic(final Player ai, final SpellAbility sa) {
+        if (ComputerUtil.preventRunAwayActivations(sa)) {
+            return false; // prevent infinite loop
+        }
+
+        return true;
+    }
+
+    /**
+     * <p>
+     * doTriggerAINoCost
+     * </p>
+     * @param sa
+     *            a {@link forge.game.spellability.SpellAbility} object.
+     * @param mandatory
+     *            a boolean.
+     *
+     * @return a boolean.
+     */
+    @Override
+    protected boolean doTriggerAINoCost(final Player ai, final SpellAbility sa, final boolean mandatory) {
+        return mandatory || checkApiLogic(ai, sa);
+    }
+
+    @Override
+    public boolean confirmAction(Player ai, SpellAbility sa, PlayerActionConfirmMode mode, String message, Map<String, Object> params) {
+        Card c = (Card)params.get("Card");
+        for (SpellAbility s : AbilityUtils.getBasicSpellsFromPlayEffect(c, ai, CardStateName.Original)) { // TODO: other states for split cards and MDFC?
+            if (!sa.matchesValidParam("ValidSA", s)) {
+                continue;
+            }
+            if (s instanceof LandAbility) {
+                // might want to run some checks here but it's rare anyway
+                return true;
```

Comment:
Discover doesn't let you play lands

Reply:
I think you need to return false or your get ClassCastException later if it encounters MDFC with land backside

Reply:
Ah, good call! Done.